### PR TITLE
feat/donation tests

### PR DESCRIPTION
**Pull request by ahmaxed**
- initial
- add github token
- feat/clean-up
- change project id
- feat/adjust action for secrets
- feat: clean up
- more clean up

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

File: cypress.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
-  "projectId": "ke77ns",
+  "projectId": "yav5fd",
```
Comment (ahmadabdolsaheb) [inline cypress.json:2]:
```suggestion
  "projectId": "ke77ns",
```

File: .gitignore [hunk 2]
```diff
@@ -208,4 +208,7 @@ client/static/mobile
 client/static/curriculum-data
 
 ### UI Components ###
-tools/ui-components/dist
+tools/ui-components/dist/*
+
+### act secrets
+my.secrets
```
Comment (ahmadabdolsaheb) [inline .gitignore:214]:
```suggestion
tools/ui-components/dist
```

File: .github/workflows/cypress-third-party.yml [hunk 3]
```diff
@@ -0,0 +1,139 @@
+name: CI - Cypress (e2e) 3rd party donation tests
```
Comment (ojeytonwilliams) [inline .github/workflows/cypress-third-party.yml:1]: Could this be made DRYer?  Maybe a composite action could be used to hold all the config that's common to this and `cypress.yml`.

@nhcarrigan do you know if composite actions would help, here?

File: cypress/integration/third-party/donate-page.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+describe('Donate page', () => {
+  it('Donation ', () => {
+    cy.exec('npm run seed');
+    cy.login();
+    cy.visit('/donate');
+    cy.wait(10000);
+    cy.get('.donation-elements').within(() => {
```
Comment (ojeytonwilliams) [inline cypress/integration/third-party/donate-page.js:7]:
```suggestion
    cy.get('.donation-elements', { timeout: 10000 }).within(() => {
```
We know what we're waiting for, so `cy.wait` is unnecessary.

File: cypress/integration/third-party/donate-page.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+      cy.fillElementsInput('cardNumber', '4242424242424242');
+      cy.fillElementsInput('cardExpiry', '1025');
+    });
+    cy.get('.confirm-donation-btn').click();
+    cy.contains('We are processing your donation.').should('be.visible');
+    cy.wait(10000);
+    cy.contains('Thank you for being a supporter.').should('be.visible');
```
Comment (ojeytonwilliams) [inline cypress/integration/third-party/donate-page.js:14]:
```suggestion
    cy.contains('Thank you for being a supporter.', { timeout: 10000 }).should(
      'be.visible'
    );
```
Same principle.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

**Comment by ahmadabdolsaheb** [comment]
Note to self:
Although the element with the class `donation-elements` is visible in the Cypress snapshots, the third party tests fail since `.donation-elements` does not seem to be passed to the `Cypress-plugin-stripe-elements` on time in the Github action.

As a result, Cypress throw the following assertion error:
`Expected to find element: undefined, but never found it.`

**Comment by ahmadabdolsaheb** [comment]
So github secrets are not being passed to pr github actions: however, they are passed once the pr is merged.

**Comment by raisedadead** [comment]
There are two ways to do this:

- You can address any pending review commits and land this on a branch other than `main` on the main repo. That is, change the PR target (base) branch to a branch that exists on the main repo.

OR

- You can test the CI on your fork by adding your keys in there & changing the setting on your fork to run workflows if it is disabled.

**Comment by ahmadabdolsaheb** [comment]
I have previously pushed it as an freeCodeCamp branch and the secrets are passed as expected; however, when this pr lands, the action will fail on all upcoming prs.

**Comment by ahmadabdolsaheb** [comment]
@raisedadead, I have previously pushed it as an freeCodeCamp branch and the secrets are passed as expected; however, when this pr lands, the action will fail on all upcoming prs.

**Comment by raisedadead** [comment]
Yes - that is correct, and this workflow will then be limited to not being available to PRs. That is we can disable the `on: pull-request` even for a while and then reactivate in a month or so when the PRs have been normalized and new PRs are up to date?

**Comment by ahmadabdolsaheb** [comment]
So since github secrets (except github token, etc) are not passed to github action workflows that are triggered by a pull request, the actions will fail regardless of when it triggered. 

An example of a secret that is not being passed to pr triggered wokflows is `secrets.CYPRESS_RECORD_KEY`

**Comment by raisedadead** [comment]
I am not sure I understand this correctly. We already have Cypress workflows, and they work on PRs. Are mistakenly confusing `on: pull_request` & `on: pull_request_target`?

**Comment by naomi-lgbt** [comment]
> We already have Cypress workflows, and they work on PRs

I believe the `CYPRESS_RECORD_KEY` only handles the video/screenshot uploads, and doesn't fail the action.

__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

**Comment by ahmadabdolsaheb** [comment]
firefox will be removed from stripe donation tests since:
```
Your project has set the configuration option: chromeWebSecurity to false

This option will not have an effect in Firefox. Tests that rely on web security being disabled will not run as expected.
```

**Comment by ahmadabdolsaheb** [comment]
I have pushed this as a branch and thing are working as expected.
https://github.com/freeCodeCamp/freeCodeCamp/actions/runs/2790419385
Let me know what you think.

**Comment by ojeytonwilliams** [comment]
> firefox will be removed from stripe donation tests...

I feel your pain.

File: .github/workflows/cypress-third-party.yml (hunk 3 above)
Comment (ojeytonwilliams) [reply .github/workflows/cypress-third-party.yml:1]: Nevermind, we can try to improve this after we've got it working on main.

File: .github/workflows/cypress-third-party.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+          record: ${{ env.CYPRESS_RECORD_KEY != 0 }}
+          start: npm run start-ci
+          wait-on: http://localhost:8000
+          wait-on-timeout: 1200
+          config: baseUrl=http://localhost:8000
+          browser: ${{ matrix.browsers }}
+          headless: true
+          spec: |
+            ${{ matrix.spec }}
+            cypress/integration/third-party/*.js
```
Comment (ojeytonwilliams) [inline .github/workflows/cypress-third-party.yml:137]:
```suggestion
          spec: cypress/integration/third-party/*.js
```
Just the third party ones, right?

File: .github/workflows/cypress-third-party.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    runs-on: ubuntu-20.04
+    needs: build-client
+    strategy:
+      fail-fast: false
+      matrix:
+        browsers: [chrome, electron]
+        node-version: [16.x]
+        include:
+          - browsers: electron
+            spec: cypress/integration/learn/challenges/projects.js
```
Comment (ojeytonwilliams) [inline .github/workflows/cypress-third-party.yml:71]:
```suggestion
```
These already get tested in the existing action.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Congrats on getting this together, @ahmadabdolsaheb, it's looking great.

I just spotted a couple of things that we don't seem to need:

**Review by ojeytonwilliams — DISMISSED** [review]
LGTM 👍

File: .github/workflows/cypress-third-party.yml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+          path: client/public/stats.json
+
+  cypress-run:
+    name: Test
+    runs-on: ubuntu-20.04
+    needs: build-client
+    strategy:
+      fail-fast: false
+      matrix:
+        browsers: [chrome, electron]
```
Comment (naomi-lgbt) [inline .github/workflows/cypress-third-party.yml:67]: Just to confirm, do we want to run the donation tests on electron?

Comment (ojeytonwilliams) [reply .github/workflows/cypress-third-party.yml:67]: Good catch, I don't think there's any need to.

Comment (ahmadabdolsaheb) [reply .github/workflows/cypress-third-party.yml]: Ok so if we are going to use Chrome only, how about removing the strategy, the artifact upload and download all-together similar to the last commit?

**Review by ojeytonwilliams — APPROVED** [review]
Sure, this seems better :+1:

**Comment by naomi-lgbt** [comment]
Just to confirm, have we resolved the secrets issue?

**Comment by ahmadabdolsaheb** [comment]
@naomi-lgbt yes. 🙏

**Comment by ahmadabdolsaheb** [comment]
Here is the last changes pushed on another branch. https://github.com/freeCodeCamp/freeCodeCamp/actions/runs/2816411977

**Review by naomi-lgbt — APPROVED** [review]
Okay, I think this should be fine to go in.

__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :ok_hand: This code looks good&nbsp;

**Pull request merged**